Add a `DataPipeline` implementation

shsms opened the pull request:
... that abstracts away the creation of the data sourcing and resampling actors,  and provides an easy way to access the logical meter or one of the pools.

Currently only provides two methods: `logical_meter` and `ev_charger_pool`.

`battery_pool` probably needs some consensus on how args should be pass and interaction between `PowerDistributingActor`,  which can be done in a subsequent PR.

This allows for the following interface:

``` python
async def run():
    await microgrid.initialize(
        host=HOST, port=PORT, resampler_config=ResamplerConfig(resampling_period_s=1.0)
    )
    grid_power = microgrid.logical_meter().grid_power()
```

This PR also renames:
1. `microgrid._microgrid` to `microgrid.connection_manager`
2. `tests/utils/mock_microgrid.py` to `mock_microgrid_client.py`

shsms committed:
Add a `DataPipeline` implementation

... that abstracts away the creation of the data sourcing and
resampling actors, and provides an easy way to access the logical
meter or one of the pools.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update `MockMicrogrid` to use `DataPipeline`

... instead of starting data pipeline actors directly.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update tests to work with new interface of `MockMicrogrid`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Was this supposed to have a different name?  Especially because the *Pools are also supposed to send control commands back to the microgrid?

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/data_pipeline.py:
```diff
@@ -0,0 +1,115 @@
+            )
+        return self._logical_meter
+
+    def ev_charger_pool(self) -> EVChargerPool:
+        """Return a reference to the EVChargerPool instance.
+
+        Returns:
+            An EVChargerPool instance.
+        """
+        if self._ev_charger_pool is None:
+            self._ev_charger_pool = EVChargerPool(
+                channel_registry=self._channel_registry,
+                resampler_subscription_sender=self._resampling_request_sender(),
+                component_ids=self._ev_charger_ids,
+            )
+        return self._ev_charger_pool
```
We might need only one instance of ChannelRegistry and many instances of EvChargerPool.
Consider to add ev_charger_ids argument to this methods (instead of in constructor): 
```
def ev_charger_pool(self, ev_charger_ids: Set[int]) -> EVChargerPool:
```
And save them by components ids.
So if user will ask for ev_pool:
* with components ids that already was created, then you should return this EvPool.
* If EVPool with this ids does not exist, then class should create this EVPool, save and return.

shsms replied:
done

ela-kotulska-frequenz commented:
> Was this supposed to have a different name? Especially because the *Pools are also supposed to send control commands back to the microgrid?

Maybe SDKInterface?

mathias-baumann-frequenz commented on src/frequenz/sdk/timeseries/data_pipeline.py:
```diff
@@ -0,0 +1,115 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Create, connect and own instances of data pipeline components."""
```
I feel this description doesn't really tell me anything about what this class can be used for

shsms replied:
I've added a few more lines,  but this is now also internal.  So end users who don't have enough context will no longer be exposed to it directly.

mathias-baumann-frequenz commented on src/frequenz/sdk/timeseries/data_pipeline.py:
```diff
@@ -0,0 +1,115 @@
+        self._logical_meter: LogicalMeter | None = None
+        self._ev_charger_pool: EVChargerPool | None = None
+
+    def _data_sourcing_request_sender(self) -> Sender[ComponentMetricRequest]:
```
I'd suggest to put all private methods to the bottom to see the public relevant methods first

shsms replied:
done.

shsms committed:
Update packages to use `connection_manager.get()`

... instead of `microgrid.get()` to get the ConnectionManager.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `microgrid._microgrid` to `microgrid.connection_manager`

This change moves the connection_manager.get() method one level
deeper, to clarify that it is a lower level interface.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update tests to access data pipeline from the microgrid module

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename tests/utils/{mock_microgrid.py => mock_microgrid_client.py}

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/__init__.py:
```diff
@@ -7,15 +7,29 @@
+
+
+async def initialize(host: str, port: int, resampler_config: ResamplerConfig) -> None:
+    """Initialize the microgrid connection manager and the data pipleine.
```
```suggestion
    """Initialize the microgrid connection manager and the data pipeline.
```

shsms replied:
wow, fixed. :-)

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -0,0 +1,207 @@
+
+
+@dataclass
+class _ActorInfo:
```
you might want to add a brief line of documentation for this class

shsms replied:
done.

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -0,0 +1,207 @@
+            )
+        return self._ev_charger_pools[key]
+
+    def _data_sourcing_request_sender(self) -> Sender[ComponentMetricRequest]:
```
documentation is missing for this function

shsms replied:
done, for the other private functions as well.

daniel-zullo-frequenz reviewed (commented):
I have only a few minor comments to check for and LGTM otherwise

I think these changes need an entry to the release notes

shsms committed:
Update RELEASE_NOTES

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
> I think these changes need an entry to the release notes

This is also done.

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -0,0 +1,224 @@
+        self._logical_meter: "LogicalMeter" | None = None
+        self._ev_charger_pools: dict[frozenset[int], "EVChargerPool"] = {}
+
+    def logical_meter(self) -> LogicalMeter:
```
I wonder if this be suitable to make it a @property

shsms replied:
I don't know, in my opinion, this does actual work, so it is not a property.  Also, it is internal and not the public api, so it doesn't matter.

daniel-zullo-frequenz replied:
I was mainly considering to use this kind of syntax `microgrid.logical_meter.grid_power()` instead but as you have said it doesn't matter as it is internal

shsms replied:
microgrid.logical_meter() is a separate function,  that calls this class method: https://github.com/frequenz-floss/frequenz-sdk-python/pull/270/files#diff-9f32f637c4ac702aa43914f3ab2949c4b6adef41b482c1d300a4262c06b1f995R191-R199

daniel-zullo-frequenz reviewed (commented):
I only have a question and LGTM otherwise
